### PR TITLE
[ADD] package_hierarchy: Constructing pacakge hierarchy from packages

### PR DESCRIPTION
**Pull request by armandcela**
Story: 1712
Signed-off-by: Armand Cela <armand.cela@unipart.io>

### CODE REVIEW

File: addons/package_hierarchy/models/stock_quant_package.py [hunk 1]
```diff
@@ -271,3 +280,36 @@ def _get_package_link_structure(self):
...
+        ordered packages that are in self recordset.
+
+        :return: recordset of package hierarchy links
+        """
+        # Setting self to sudo to over pass access rights at this moment, have to be reviewed later
+        self = self.sudo()
+        PackageHierarchyLink = self.env["package.hierarchy.link"]
+
+        # Empty recordset
+        package_hierarchies = self.env["package.hierarchy.link"]
```
Comment (ws-mps) [inline addons/package_hierarchy/models/stock_quant_package.py:295]: Shouldn't this be `PackageHierarchyLink.browse()` ?

Comment (armandcela) [reply addons/package_hierarchy/models/stock_quant_package.py:295]: It is the same thing in fact, but yeah due to what have been discussed about coding standards happy to make that change

File: addons/package_hierarchy/models/stock_quant_package.py [hunk 2]
```diff
@@ -271,3 +280,36 @@ def _get_package_link_structure(self):
...
+                [("child_id", "=", package.id), ("parent_id", "=", parent_package.id)], limit=1
+            )
+            if not hierarchy_link:
+                hierarchy_link = PackageHierarchyLink.create({
+                    "child_id": package.id,
+                    "parent_id": parent_package.id
+                })
+            package_hierarchies |= hierarchy_link
+        hierarchy_link = PackageHierarchyLink.search(
+                [("child_id", "=", parent_package.id), ("parent_id", "=", False)], limit=1
```
Comment (ws-mps) [inline addons/package_hierarchy/models/stock_quant_package.py:308]: Will parent_package have value if there is only one element in self?

File: addons/package_hierarchy/models/stock_quant_package.py [hunk 3]
```diff
@@ -10,6 +10,15 @@
 _logger = logging.getLogger(__name__)
 
 
+def pairwise(original_list):
```
Comment (ws-mps) [inline addons/package_hierarchy/models/stock_quant_package.py:13]: Doesn't itertools already have this method?

File: addons/package_hierarchy/models/stock_quant_package.py (hunk 2 above)
Comment (armandcela) [reply addons/package_hierarchy/models/stock_quant_package.py:308]: Very good spot didnt thought about it :(
Will check , most probably will raise error i think :)

**Review by ws-mps — APPROVED** [review]

**Comment by LoKKeR96** [comment]
Happy but I would want test.

**Pull request merged**